### DISCUSSION TITLE
Windows support? (EDIT: macOS et al too)

### DISCUSSION THREAD

**Discussion by tez011** [Flow-IPC/ipc#98] [Ideas]
How separated is the Linux-specific code from the rest of flow-ipc? Wondering how feasible it would be to make a demo of this run on windows.

**Comment by ygoldfeld** [comment]
That's a good question - generally speaking there are no major modules that are written specifically for Linux or *nix; it is more a matter of certain specifics which happen to be Linux-specific. For example, /dev/shm might be used to list known SHM pools; and x86-64 pointer tagging is used in a certain spot.

I'll answer here like this:
- Every unportable thing *is* protected by an `#if FLOW_OS_LINUX` (and a few similar) and either `#error`s out or `static_assert(false)` if not the case. At least a strong effort was made to do this always.
- I recently actually went through the exercise of looking for *all* such things across the code and figuring out what it would take to port these details to:
  - Primary: macOS (as opposed to only Linux at the moment).
  - Primary: ARM64 (as opposed to only x86-64). (There is only one thing that's specifically x86-64-reliant: a pointer-tagging technique.) That is because at least modern Macs are on ARM64.
  - Secondary: BSD variants (FreeBSD, OpenBSD).
  - Tertiary: Windows.
  - The primary/secondary/tertiary ranking might be wrong; depends on demand.
- I have these notes, but I haven't published them yet. But I will soon! When I do I'll put them in an Issue probably and link to that in here. With those how-to notes available, it should be very manageable and incrementally doable.
- The overall impression is: it is probably a couple weeks of work for macOS+ARM64, building out some replacement for /dev/shm and /dev/mqueue being really the only actual algorithmic challenge. The ARM64 difference is really trivial.
- Similarly for BSDs, as long as it's easy enough to test it CI-wise (this part I have not researched). Actually it might be easier, with most changes for Mac applying here.
- Similarly for Windows, as long as it's easy enough to test it CI-wise (this part I have not researched). We're not Windows-y people, so the expertise from a Windows-y person would help, but who knows?
  - There might be another layer here which I did *not* explore at all: the CI pipeline is pretty Unix-oriented. So this might take some research on our parts... but PRs welcome! Whereas macOS + BSD are still Unix-y, so the CI pipeline - if it can be executed on the proper hardware - will still work.

At the moment I owe those notes (soon).

**Reply by ygoldfeld** [reply]
Here's the ticket with all the info!

Flow-IPC/ipc#101
